runtime/src/iree/base/status.c

```diff
@@ -1150,6 +1139,144 @@ IREE_API_EXPORT bool iree_status_format(iree_status_t status,
+          status, /*buffer_capacity=*/0,
+          /*buffer=*/NULL, &message_buffer_size, /*has_prefix=*/false))) {
+    iree_status_free(status);
+    return iree_status_from_code(code);
+  }
+  message_buffer_size++;  // NUL
+
+#if (IREE_STATUS_FEATURES & IREE_STATUS_FEATURE_SOURCE_LOCATION) != 0
+  // Grab storage for the source file location.
+  iree_status_storage_t* storage = iree_status_storage(status);
+  const char* file = NULL;
+  uint32_t line = 0;
+  if (storage) {
+    file = storage->file;
+    line = storage->line;
+  }
+#endif  // has IREE_STATUS_FEATURE_SOURCE_LOCATION
+
+  // Allocate storage with the additional room to store the formatted message
+  // and the source file path if present.
+  size_t unaligned_storage_size =
+      sizeof(iree_status_storage_t) + message_buffer_size;
+#if (IREE_STATUS_FEATURES & IREE_STATUS_FEATURE_SOURCE_LOCATION) != 0
+  size_t file_storage_size = file ? strlen(file) + 1 : 0;
+  unaligned_storage_size += file_storage_size;
+#endif
+  size_t storage_alignment = (IREE_STATUS_CODE_MASK + 1);
+  size_t storage_size =
+      iree_host_align(unaligned_storage_size, storage_alignment);
+  iree_status_storage_t* new_storage =
+      (iree_status_storage_t*)iree_aligned_alloc(storage_alignment,
+                                                 storage_size);
+  if (IREE_UNLIKELY(!new_storage)) {
+    iree_status_free(status);
+    return iree_status_from_code(code);
+  }
+  memset(new_storage, 0, sizeof(*new_storage));
+
+  char* message_data = (char*)new_storage + sizeof(iree_status_storage_t);
+#if (IREE_STATUS_FEATURES & IREE_STATUS_FEATURE_SOURCE_LOCATION) != 0
+  message_data += file_storage_size;
+#endif  // has IREE_STATUS_FEATURE_SOURCE_LOCATION
+  size_t res_length;
+  // Format the status message directly into the region allocated for it.
+  bool ret =
+      iree_status_format_message(status, message_buffer_size, message_data,
+                                 &res_length, /*has_prefix=*/false);
+  new_storage->message.size = message_buffer_size - 1;
+  new_storage->message.data =
+      (const char*)new_storage + sizeof(iree_status_storage_t);
+  iree_status_t new_status =
+      (iree_status_t)((uintptr_t)new_storage | (code & IREE_STATUS_CODE_MASK));
+
+  if (IREE_UNLIKELY(!ret)) {
+    iree_status_free(new_status);
+    iree_status_free(status);
+    return iree_status_from_code(code);
+  }
+
+#if (IREE_STATUS_FEATURES & IREE_STATUS_FEATURE_SOURCE_LOCATION) != 0
+  if (file) {
+    new_storage->file = storage->file;
+    char* storage_file = (char*)new_storage + sizeof(iree_status_storage_t);
+    // Copy the file into the storage allocated for it.
+    memcpy(storage_file, file, file_storage_size);
+    new_storage->file = (const char*)storage_file;
+    new_storage->message.data += file_storage_size;
+  }
+  new_storage->line = line;
+#endif  // has IREE_STATUS_FEATURE_SOURCE_LOCATION
+
+  iree_status_free(status);
+  return new_status;
+#endif
```

Comment:
```suggestion
#endif  // has any IREE_STATUS_FEATURES
```

for readability it's also sometimes nicer when the function being #ifdefed out is beyond a page to instead just define the function twice - one in the #if as a no-op and the other in the #else - then no need for a reader to track what parts of a function are enabled and instead can easily see the function is just a return _or_ actual work